internal/controller/core/device_controller.go

```diff
@@ -133,29 +127,46 @@ func (r *DeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ c
 		})
 		obj.Status.Phase = v1alpha1.DevicePhaseProvisioning
 		r.Recorder.Event(obj, "Normal", "ProvisioningStarted", "Device provisioning has started")
-		// TODO(swagner-de): Start POAP Process.
-		_ = tmpl // <-- Use the template.
-		return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+		return ctrl.Result{}, nil
 
 	case v1alpha1.DevicePhaseProvisioning:
-		log.Info("Device is in provisioning phase, checking completion")
-		// TODO(swagner-de): Check if POAP Process is complete.
-		ready := true // <-- This should be replaced with actual readiness check logic.
-		if !ready {
-			// If the device is not ready yet, we requeue the request to check again later.
+		activeProv := obj.GetActiveProvisioning()
+		if activeProv == nil {
+			log.Info("Device has not made a provisioning request yet")
+			return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+		}
+		if activeProv.StartTime.Add(time.Hour).Before(time.Now()) {
+			obj.Status.Phase = v1alpha1.DevicePhaseFailed
+			r.Recorder.Event(obj, "Warning", "ProvisioningFailed", "Device provisioning has timed out")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: 20 * time.Minute}, nil
+
+	case v1alpha1.DevicePhaseProvisioningCompleted:
+		// we will finalize the provisioning here, either wait for the reboot to be initiated/completed
+		// or run post-provisioning checks instantly if no reboot is required
+		activeProv := obj.GetActiveProvisioning()
+		if activeProv == nil {
+			err := errors.New("device went into provisioning completed phase, but no active provisioning found")
+			log.Error(err, "Failed to finalize provisioning")
+			return ctrl.Result{}, err
```

Comment:
```suggestion
			return ctrl.Result{}, reconcile.TerminalError(err)
```
We should use a [reconcile.TerminalError](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#TerminalError) as there is no point in retrying on this error.